contxt/auth/cli.py

```diff
@@ -1,16 +1,72 @@
-from getpass import getpass
-from json import dump, load
+import time
+import webbrowser
+from json import dump, load, loads
 from pathlib import Path
 from typing import Any, Dict, Optional
 
 from auth0.v3.authentication import GetToken
+from requests.exceptions import HTTPError
+
+from contxt.services.api import Api
 
 from ..services.auth import AuthService
 from ..utils import make_logger
 from . import Auth, Token, TokenProvider
 
 logger = make_logger(__name__)
 
+CLI_CLIENT_ID = "bleED0RUwb7CJ9j7D48tqSiSZRZn29AV"
+
+
+class DeviceAuthPendingException(Exception):
+    pass
+
+
+class DeviceAuthTimeout(Exception):
+    pass
+
+
+class DeviceAuthDenied(Exception):
+    pass
+
+
+class Auth0DeviceProvider(Api):
+    def __init__(self, auth0_tenant):
+        self.base_url = auth0_tenant
+        self.auth_service = AuthService()
+        super().__init__(base_url=f"https://{self.base_url}")
+
+    def get_device_code_url(self):
+        data = {
+            "client_id": CLI_CLIENT_ID,
+            "scope": "offline_access",
+            "audience": self.auth_service.client_id,
+        }
+
+        return self.post("oauth/device/code", data)
+
+    def get_access_token(self, device_code):
+        data = {
+            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+            "device_code": device_code,
+            "client_id": CLI_CLIENT_ID,
+        }
+
+        try:
+            resp = self.post("oauth/token", data)
+        except HTTPError as e:
+            resp = loads(e.response.content)
```

Comment:
```suggestion
            resp = e.response.json()
```

also, you seem to be raising exceptions here which just get caught and re-raised later. if you wanted to simplify a bit:
```python
            if e.response.json()["error"] != "authorization_pending":
                raise e
            resp = None # we are in a pending state, return None rather than raising a hard exception
```

Reply:
I agree with not raising an exception when device auth is pending. It should either be swallowed or replaced with a friendlier `print("Waiting for Browser Authorization")` or some other equivalent message.